core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -487,6 +487,18 @@ public static SqlCall stripSeparator(SqlCall call) {
         -1);
   };
 
+  /**
+   * Returns the element type of a ARRAY or MULTISET.
+   *
+   * <p>For example, given <code>INTEGER ARRAY or MULTISET ARRAY</code>, returns
+   * <code>INTEGER</code>.
+   */
+  public static final SqlReturnTypeInference TO_COLLECTION_ELEMENT =
+      ARG0.andThen(SqlTypeTransforms.TO_COLLECTION_ELEMENT_TYPE);
```

Comment:
Will fail with an empty array:
https://github.com/apache/calcite/pull/3218#discussion_r1209622163

Reply:
@MasseGuillaume  
it doesn't faill, because array() type is UNKNOWN NOT NULL ARRAY NOT NULL,
so the TO_COLLECTION_ELEMENT will return UNKNOWN NOT NULL